schema/deploy/warehouse/identifier/triggers/barcode-distance-check.sql

```diff
@@ -31,8 +37,9 @@ create or replace function warehouse.identifier_barcode_distance_check() returns
 
         select barcode into conflicting_barcode from (
             select barcode
-              from warehouse.identifier
-             where hamming_distance_lte(lower(barcode), new_barcode, minimum_distance-1) < minimum_distance
+            from warehouse.identifier
+            where slices && NEW.slices
```

Comment:
If `slices && NEW.slices` returns null because one of them is null, I think this query fails in the wrong direction (potentially doesn't compare distance to anything). Seems like it might be best to guard against that?

Reply:
True true. I've added the null checks in the fixup commit.

Reply:
See what you think about what I've done in the 'move to index' fixup commit. I'm handling the case where barcode_slices(barcode) is an empty array. When I looked for `barcode_slices(barcode) is null` this caused pg to not use the index.
The barcode_slices function returns null if the input text is null. If the barcode is null in an identifier row we don't need to check the hamming distance against it.

Reply:
Hmm. So the new condition is:

```
(barcode_slices(barcode) = '{}' or barcode_slices(barcode) && barcode_slices(NEW.barcode))
```

but I don't think this covers the null edge case I raised.

1. It doesn't catch cases where `barcode_slices(NEW.barcode)` is null, which could happen for a variety of reasons. The whole expression would evaluate as null in that case and thus skip the distance calc.

2. The `barcode_slices()` function never returns an empty array: it either returns null or an array with values. So `barcode_slices(barcode) = '{}'` will always be false.

My suggestion is to do this instead:

```
barcode_slices(barcode) && barcode_slices(NEW.barcode) is not false
```

which will test Hamming distance if the slices overlap or if (for whatever reason) either the LHS or RHS is null. I expect this still uses the index, but it would be good to double check that empirically.

Looking at this again right now, I'm also realizing that `barcode_slices()` likely wants to be case-insensitive as well, otherwise barcodes may be inappropriately skipped due to case differences alone. My suggestion is to switch the input and output types of `barcode_slices()` from `text` to `citext`, as the function is specifically intended for `warehouse.identifier.barcode` which is `citext`. (There are perf. reasons why the Hamming distance functions don't use `citext`, but since `barcode_slices()` is pre-computed by the index, those same reasons don't apply.)

Reply:
I've changed the function to take/return citext. I also renamed the index to end in "_idx" as is customary.

I couldn't come up with a way to make `barcode_slices(barcode) && barcode_slices(NEW.barcode) is not false` use the index. What I did instead was to make two code paths .... if the slices on the new barcode is null don't use the intersection shortcut; if slices on the new barcode is not null do use the shortcut. This means duplicating most of the logic, but it lets the happy path use the index. It also lets us add other conditions to the unhappy path if we need them.

This does not cover the case when barcode_slices(barcode) is null for some rows in the table. This would happen if the barcode column itself is null. In that case it's fine not to compare against it. 
Do you see other scenarios when barcode_slices(barcode) could be null?

Reply:
Meh, I don't like what I added. I couldn't use what you suggested `barcode_slices(barcode) && barcode_slices(NEW.barcode) is not false` because that doesn't use the index, so I tried to cover one scenario. That seems like a lot of copied code for little value.

So, what are the scenarios that would lead to null slices? If the barcode is null then the slices will be null for sure. What other scenarios are we wanting to look for?

Reply:
Ah, thanks for testing it out. Sorry the suggestion wasn't workable. I think a "happy path" and "grudging path" are ok, but I agree it'd be nice to avoid if we can since duplicate code paths allow for the possibility of bugs due to drift, which is bad for something like unique identifier control. On the other hand, it's nice to cover all edge cases in important code like this so that they can't bite us later.

`barcode_slices()` returns null if the input text is a) null or b) shorter than the slice width (currently 2). (a) should be prevented by the `not null` constraint on `warehouse.identifier.barcode` and (b) should be prevented by the `length = 8` check constraint on `warehouse.identifier.barcode`. But as both of those constraints are independent of the distance check (e.g. could be removed without the distance check knowing), it seems most safe for the distance check to not rely on them. However, this may not be the most pragmatic choice given other considerations... but also read on.

One option to avoid the duplication but also handle the null edge cases is to revert the `if new_slices is null` conditional and change the query criteria in the trigger to

```
-- Use barcode slices to reduce the comparison set for pairwise
-- distance calculations.  If slices can't be computed/are missing for
-- either side, then always check the distance out of an abundance of
-- precaution.
(barcode_slices(barcode) && new_slices or barcode_slices(barcode) is null or new_slices is null)
```

The `is null` check can then be made fast by adding an additional btree index, e.g.:

```
create index identifier_barcode_slices_btree_idx on warehouse.identifier using btree (barcode_slices(barcode));
```

The new index will take up additional space, of course, but on my local instance it's only a tiny ~64MB for the whole thing. To reduce the size, we could even further target the index on the `is null` expression:

```
create index identifier_barcode_slices_is_null_idx on warehouse.identifier using btree ((barcode_slices(barcode) is null));
```

which is only ~10MB.

<details><summary>Example query plan (with the first index above)</summary>

```
2020-12-08 12:35:49.247 PST [27379] tom@seattleflu LOG:  duration: 54.470 ms  plan:
        Query Text: select barcode                          from (
                    select barcode
                      from warehouse.identifier
                     where (barcode_slices(barcode) && new_slices or barcode_slices(barcode) is null or new_slices is null)
                       and hamming_distance_lte(lower(barcode), new_barcode, minimum_distance-1) < minimum_distance
                    except
                    select old_barcode
                     limit 1
                ) x  
        Subquery Scan on x  (cost=1372.32..1374.05 rows=1 width=32)
          ->  Limit  (cost=1372.32..1374.04 rows=1 width=36)
                ->  HashSetOp Except  (cost=1372.32..12061.48 rows=6225 width=36)
                      ->  Append  (cost=1372.32..12045.92 rows=6226 width=36)
                            ->  Subquery Scan on "*SELECT* 1"  (cost=1372.32..12014.77 rows=6225 width=13)
                                  ->  Gather  (cost=1372.32..11952.52 rows=6225 width=9)
                                        Workers Planned: 2
                                        ->  Parallel Bitmap Heap Scan on identifier  (cost=372.32..10330.02 rows=2594 width=9)
                                              Recheck Cond: ((barcode_slices(barcode) && '{1__a5,2__5d,3__de,4__e7,5__79,6__9e,7__ef}'::citext[]) OR (barcode_slices(barcode) IS NULL))
                                              Filter: (hamming_distance_lte(lower((barcode)::text), 'a5de79ef'::text, 2) < 3)
                                              ->  BitmapOr  (cost=372.32..372.32 rows=18755 width=0)
                                                    ->  Bitmap Index Scan on identifier_barcode_slices_idx  (cost=0.00..182.85 rows=16380 width=0)
                                                          Index Cond: (barcode_slices(barcode) && '{1__a5,2__5d,3__de,4__e7,5__79,6__9e,7__ef}'::citext[])
                                                    ->  Bitmap Index Scan on __tmp_test  (cost=0.00..186.36 rows=2375 width=0)
                                                          Index Cond: (barcode_slices(barcode) IS NULL)
                            ->  Subquery Scan on "*SELECT* 2"  (cost=0.00..0.02 rows=1 width=36)
                                  ->  Result  (cost=0.00..0.01 rows=1 width=32)
```

</details>

Reply:
@tsibley I went with adding the smaller index to capture whether the slices is null.
With that index created, all was working great except for the `or new_slices is null` part. With that check, the operation was slow again. Ah, it looks like PG has issues with `limit 1`. See https://stackoverflow.com/questions/21385555/postgresql-query-very-slow-with-limit-1/27237698 and maybe https://www.postgresql.org/message-id/flat/CA%2BU5nMLbXfUT9cWDHJ3tpxjC3bTWqizBKqTwDgzebCB5bAGCgg%40mail.gmail.com.

I ended up reverting the change we made a few weeks ago that added `limit 1`. By reverted I mean that I copy and pasted from an earlier commit that had this in there. I see it working correctly:
```ERROR:  INSERT of (barcode='5220aa94') excluded by minimum substitution distance check
DETAIL:  Barcode '5220aa94' is closer than 3 substitutions away from existing barcodes {5b20aa97,5626aa94,52b03a94,522f1a94}
HINT:  If the INSERT relied on default values, retrying it will use a new value and possibly succeed
CONTEXT:  PL/pgSQL function warehouse.identifier_barcode_distance_check() line 38 at RAISE
SQL state: 23P01```

I've updated the commit message with all the changes and the results from my runs today.
I think we're looking good with those two null checks in there working fast.

Reply:
Thanks for digging into this. The pathologic query planner behaviour possible around small limits not using indexes is really interesting and useful to know about.